sklearn/ensemble/voting_classifier.py

```diff
@@ -0,0 +1,189 @@
+      the sums of the predicted probalities.
+
+    weights : array-like, shape = [n_classifiers], optional (default=`None`)
+      Sequence of weights (`float` or `int`) to weight the occurances of
+      predicted class labels (`hard` voting) or class probabilities
+      before averaging (`soft` voting). Uses uniform weights if `None`.
+
+    Attributes
+    ----------
+    classes_ : array-like, shape = [n_predictions]
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.linear_model import LogisticRegression
+    >>> from sklearn.naive_bayes import GaussianNB
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> clf1 = LogisticRegression(random_state=1)
+    >>> clf2 = RandomForestClassifier(random_state=1)
+    >>> clf3 = GaussianNB()
+    >>> X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
+    >>> y = np.array([1, 1, 1, 2, 2, 2])
+    >>> eclf1 = VotingClassifier(estimators=[clf1, clf2, clf3], voting='hard')
+    >>> eclf1 = eclf1.fit(X, y)
+    >>> print(eclf1.predict(X))
+    [1 1 1 2 2 2]
+    >>> eclf2 = VotingClassifier(estimators=[clf1, clf2, clf3], voting='soft')
+    >>> eclf2 = eclf2.fit(X, y)
+    >>> print(eclf2.predict(X))
+    [1 1 1 2 2 2]
+    >>> eclf3 = VotingClassifier(estimators=[clf1, clf2, clf3],
+    ...                          voting='soft', weights=[2,1,1])
+    >>> eclf3 = eclf3.fit(X, y)
+    >>> print(eclf3.predict(X))
+    [1 1 1 2 2 2]
+    >>>
+    """
+    
+    def __init__(self, estimators, voting='hard', weights=None):
+        self.estimators = estimators
+        self.voting = voting
+        self.weights = weights
+
+    def fit(self, X, y):
+        """ Fit the estimators.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+        if isinstance(y, np.ndarray) and len(y.shape) > 1 and y.shape[1] > 1:
+            raise NotImplementedError('Multilabel classification'
+                                      ' not implemented, yet.')
+
+        if self.voting not in ('soft', 'hard'):
+            raise ValueError("Voting must be 'soft' or 'hard'; got (voting=%r)"
+                             % voting)
+
+        if self.weights and len(self.weights) != len(self.estimators):
+            raise ValueError('Number of classifiers and weights must be equal'
+                             '; got %d weights, %d estimators'
+                             % (len(self.weights), len(self.estimators)))
+
+        self.le_ = LabelEncoder()
+        self.le_.fit(y)
+        self.classes_ = self.le_.classes_
+
+        for clf in self.estimators:
+            clf.fit(X, self.le_.transform(y))      
```

Comment:
Calling fit on a model should never mutate the constructor parameter state by default. I think we should rewrite this as:

```
self.estimators_ = []
for clf in self.estimators:
    fitted_clf = clone(clf).fit(X, self.le_.transform(y)) 
    self.estimators.append(fitted_clf)
```

Then in the  `_predict` and `_predict_proba` we should use `self.estimators_` instead of `self.estimators`. The newly introduced `estimators_` attribute should be documented in the class docstring.

Also add a `check_fitted(self, 'estimators_')` in the `_predict` and `_predict_proba` methods.

@amueller do you think we should try to write a `test_common.py` to detect those kinds of problems?